kernel/driver.cc

```diff
@@ -244,11 +244,35 @@ int main(int argc, char **argv)
 	bool mode_q = false;
 
 #if defined(YOSYS_ENABLE_READLINE) || defined(YOSYS_ENABLE_EDITLINE)
-	if (getenv("HOME") != NULL) {
-		yosys_history_file = stringf("%s/.yosys_history", getenv("HOME"));
-		read_history(yosys_history_file.c_str());
-		yosys_history_offset = where_history();
-	}
+    std::string state_dir;
+    #if defined(_WIN32)
+        if (getenv("HOMEDRIVE") != NULL && getenv("HOMEPATH") != NULL) {
+            state_dir = stringf("%s%s/.local/state", getenv("HOMEDRIVE"), getenv("HOMEPATH"));
+        } else {
+            log("$HOMEDRIVE and/or $HOMEPATH is empty. No history file will be created.");
+        }
+    #else
+        if (getenv("XDG_STATE_HOME") == NULL || getenv("XDG_STATE_HOME")[0] == '\0') {
+            if (getenv("HOME") != NULL) {
+                state_dir = stringf("%s/.local/state", getenv("HOME"));
+            } else {
+                log("$HOME is empty. No history file will be created.");
+            }
+        } else {
+            state_dir = stringf("%s", getenv("XDG_STATE_HOME"));
+        }
+    #endif
+
+    if (!state_dir.empty()) {
+        std::string yosys_dir = state_dir + "/yosys";
+        create_directory(yosys_dir);
+
+        yosys_history_file = yosys_dir + "/history";
+        read_history(yosys_history_file.c_str());
+        yosys_history_offset = where_history();
+    } else {
+        log("state_dir is empty. No history file will be created.");
```

Comment:
Maybe print what `state_dir` is? As a downstream user of Yosys I would have zero clue what this means.